src/libraries/System.Private.CoreLib/src/System/DateOnly.cs

```diff
@@ -0,0 +1,904 @@
+        /// <returns>true if s was converted successfully; otherwise, false.</returns>
+        public static bool TryParseExact(ReadOnlySpan<char> s, ReadOnlySpan<char> format, out DateOnly result) => TryParseExact(s, format, DateTimeFormatInfo.CurrentInfo, DateTimeStyles.None, out result);
+
+        /// <summary>
+        /// Converts the specified span representation of a date to its DateOnly equivalent using the specified format, culture-specific format information, and style.
+        /// The format of the string representation must match the specified format exactly. The method returns a value that indicates whether the conversion succeeded.
+        /// </summary>
+        /// <param name="s">A span containing the characters representing a date to convert.</param>
+        /// <param name="format">The required format of s.</param>
+        /// <param name="provider">An object that supplies culture-specific formatting information about s.</param>
+        /// <param name="style">A bitwise combination of one or more enumeration values that indicate the permitted format of s.</param>
+        /// <param name="result">When this method returns, contains the DateOnly value equivalent to the date contained in s, if the conversion succeeded, or MinValue if the conversion failed. The conversion fails if the s is empty string, or does not contain a date that correspond to the pattern specified in format. This parameter is passed uninitialized.</param>
+        /// <returns>true if s was converted successfully; otherwise, false.</returns>
+        public static bool TryParseExact(ReadOnlySpan<char> s, ReadOnlySpan<char> format, IFormatProvider? provider, DateTimeStyles style, out DateOnly result)
+        {
+            result = default;
+            if ((style & ~DateTimeStyles.AllowWhiteSpaces) != 0)
+            {
+                return false;
+            }
+
+            if (format.Length == 1)
+            {
+                if (format[0] == 'o' || format[0] == 'O')
+                {
+                    format = OFormat;
+                    provider = CultureInfo.InvariantCulture.DateTimeFormat;
+                }
+                else if (format[0] == 'r' || format[0] == 'R')
+                {
+                    format = RFormat;
+                    provider = CultureInfo.InvariantCulture.DateTimeFormat;
+                }
+            }
+
+            DateTimeResult dtResult = default;
+            dtResult.Init(s);
+
+            if (!DateTimeParse.TryParseExact(s, format, DateTimeFormatInfo.GetInstance(provider), style, ref dtResult))
+            {
+                return false;
+            }
+
+            if ((dtResult.flags & ParseFlagsDateMask) != 0)
+            {
+                return false;
+            }
+
+            result = new DateOnly(DayNumberFromDateTime(dtResult.parsedDate));
+            return true;
+        }
```

Comment:
It'd be nice to structure this in a way where we don't need to write to the `out` multiple times in the success path, e.g.
```C#
            if ((style & ~DateTimeStyles.AllowWhiteSpaces) == 0)
            {
                if (format.Length == 1)
                {
                    switch (format[0])
                    {
                        case 'o': case 'O':
                            format = OFormat;
                            provider = CultureInfo.InvariantCulture.DateTimeFormat;
                            break;

                        case 'r': case 'R':
                            format = RFormat;
                            provider = CultureInfo.InvariantCulture.DateTimeFormat;
                            break;
                    }
                }

                DateTimeResult dtResult = default;
                dtResult.Init(s);

                if (DateTimeParse.TryParseExact(s, format, DateTimeFormatInfo.GetInstance(provider), style, ref dtResult) &&
                    (dtResult.flags & ParseFlagsDateMask) == 0)
                {
                    result = new DateOnly(DayNumberFromDateTime(dtResult.parsedDate));
                    return true;
                }
            }

            result = default;
            return false;
```

Reply:
will do